File: app/src/main/java/me/devsaki/hentoid/parsers/content/TsuminoContent.java
package me.devsaki.hentoid.parsers.content;

import static me.devsaki.hentoid.enums.Site.TSUMINO;

import androidx.annotation.NonNull;

import org.jsoup.nodes.Element;

import java.util.Collections;
import java.util.List;

import javax.annotation.Nonnull;

import me.devsaki.hentoid.database.domains.AttributeMap;
import me.devsaki.hentoid.database.domains.Content;
import me.devsaki.hentoid.enums.AttributeType;
import me.devsaki.hentoid.enums.StatusContent;
import me.devsaki.hentoid.parsers.ParseHelper;
import me.devsaki.hentoid.util.Helper;
import me.devsaki.hentoid.util.StringHelper;
import pl.droidsonroids.jspoon.annotation.Selector;

public class TsuminoContent extends BaseContentParser {
    @Selector(value = "div.book-page-cover a", attr = "href", defValue = "")
    private String galleryUrl;
    @Selector(value = "img.book-page-image")
    private Element cover;
    @Selector(value = "div#Title", defValue = "")
    private String title;
    @Selector(value = "div#Uploaded", defValue = "")
    private String uploadDate;
    @Selector(value = "div#Pages", defValue = "")
    private String pages;
    @Selector(value = "div#Artist a")
    private List<Element> artists;
    @Selector(value = "div#Group a")
    private List<Element> circles;
    @Selector(value = "div#Tag a")
    private List<Element> tags;
    @Selector(value = "div#Parody a")
    private List<Element> series;
    @Selector(value = "div#Character a")
    private List<Element> characters;
    @Selector(value = "div#Category a")
    private List<Element> categories;


    public Content update(@NonNull final Content content, @Nonnull String url, boolean updateImages) {
        content.setSite(TSUMINO);
        String theUrl = galleryUrl.isEmpty() ? url : galleryUrl;
        if (theUrl.isEmpty()) return new Content().setStatus(StatusContent.IGNORED);

        content.setUrl(theUrl.replace("/Read/Index", ""));
        String coverUrl = (cover != null) ? ParseHelper.getImgSrc(cover) : "";
        if (!coverUrl.startsWith("http")) coverUrl = TSUMINO.getUrl() + coverUrl;
        content.setCoverImageUrl(coverUrl);
        content.setTitle(StringHelper.removeNonPrintableChars(title));

        content.setUploadDate(Helper.parseDateToEpoch(uploadDate, "yyyy MMMM dd")); // e.g. 2021 December 13

        AttributeMap attributes = new AttributeMap();
        ParseHelper.parseAttributes(attributes, AttributeType.ARTIST, artists, false, TSUMINO);
        ParseHelper.parseAttributes(attributes, AttributeType.CIRCLE, circles, false, TSUMINO);
        ParseHelper.parseAttributes(attributes, AttributeType.TAG, tags, false, TSUMINO);
        ParseHelper.parseAttributes(attributes, AttributeType.SERIE, series, false, TSUMINO);
        ParseHelper.parseAttributes(attributes, AttributeType.CHARACTER, characters, false, TSUMINO);
        ParseHelper.parseAttributes(attributes, AttributeType.CATEGORY, categories, false, TSUMINO);
        content.putAttributes(attributes);

        if (updateImages) {
            content.setImageFiles(Collections.emptyList());
            content.setQtyPages((pages.length() > 0) ? Integer.parseInt(pages) : 0);
        }

        return content;
    }
}

File: app/src/main/java/me/devsaki/hentoid/parsers/content/TsuminoContent.kt
package me.devsaki.hentoid.parsers.content

import me.devsaki.hentoid.database.domains.AttributeMap
import me.devsaki.hentoid.database.domains.Content
import me.devsaki.hentoid.enums.AttributeType
import me.devsaki.hentoid.enums.Site
import me.devsaki.hentoid.enums.StatusContent
import me.devsaki.hentoid.parsers.cleanup
import me.devsaki.hentoid.parsers.getImgSrc
import me.devsaki.hentoid.parsers.parseAttributes
import me.devsaki.hentoid.util.Helper
import org.jsoup.nodes.Element
import pl.droidsonroids.jspoon.annotation.Selector

class TsuminoContent : BaseContentParser() {
    @Selector(value = "div.book-page-cover a", attr = "href", defValue = "")
    private lateinit var galleryUrl: String

    @Selector(value = "img.book-page-image")
    private var cover: Element? = null

    @Selector(value = "head [property=og:title]", attr = "content", defValue = "")
    private lateinit var title: String

    @Selector(value = "div#Uploaded", defValue = "")
    private lateinit var uploadDate: String

    @Selector(value = "div#Pages", defValue = "")
    private lateinit var pages: String

    @Selector(value = "div#Artist a")
    private var artists: List<Element>? = null

    @Selector(value = "div#Group a")
    private var circles: List<Element>? = null

    @Selector(value = "div#Tag a")
    private var tags: List<Element>? = null

    @Selector(value = "div#Parody a")
    private var series: List<Element>? = null

    @Selector(value = "div#Character a")
    private var characters: List<Element>? = null

    @Selector(value = "div#Category a")
    private var categories: List<Element>? = null


    override fun update(content: Content, url: String, updateImages: Boolean): Content {
        content.setSite(Site.TSUMINO)
        val theUrl = galleryUrl.ifEmpty { url }
        if (theUrl.isEmpty()) return Content().setStatus(StatusContent.IGNORED)
        content.setRawUrl(theUrl)

        var coverUrl = ""
        cover?.let {
            coverUrl = getImgSrc(it)
        }
        if (!coverUrl.startsWith("http")) coverUrl = Site.TSUMINO.url + coverUrl
        content.setCoverImageUrl(coverUrl)
        content.setTitle(cleanup(title))
        content.setUploadDate(
            Helper.parseDateToEpoch(uploadDate, "yyyy MMMM dd")
        ) // e.g. 2021 December 13
        val attributes = AttributeMap()
        parseAttributes(attributes, AttributeType.ARTIST, artists, false, Site.TSUMINO)
        parseAttributes(attributes, AttributeType.CIRCLE, circles, false, Site.TSUMINO)
        parseAttributes(attributes, AttributeType.TAG, tags, false, Site.TSUMINO)
        parseAttributes(attributes, AttributeType.SERIE, series, false, Site.TSUMINO)
        parseAttributes(
            attributes,
            AttributeType.CHARACTER,
            characters,
            false,
            Site.TSUMINO
        )
        parseAttributes(
            attributes,
            AttributeType.CATEGORY,
            categories,
            false,
            Site.TSUMINO
        )
        content.putAttributes(attributes)
        if (updateImages) {
            content.setImageFiles(emptyList())
            content.setQtyPages(if (pages.isNotEmpty()) pages.toInt() else 0)
        }
        return content
    }
}